### PR TITLE
Add OP_DIGEST

### PR DESCRIPTION
**Pull request by msinkec**
Implements `OP_DIGEST` (`0xc3` / 195), a generic hash opcode that pops `<data> <hash_type>` and pushes the digest of `data` under the selected algorithm. The selector is the top stack item, so scripts read as `<data> <hash_type> OP_DIGEST`.

Supported hash types:
- `1` = SHA-256
- `2` = SHA-1
- `3` = RIPEMD-160
- `4` = Keccak-256, legacy/Ethereum Keccak, distinct from NIST SHA3
- `5` = SHA3-256, NIST FIPS-202

Unknown selectors fail the script. Selector parsing uses normal script number handling, including minimal-encoding and size checks.

### Changes
- `opcode.go`: renames the free `0xc3` slot to `OP_DIGEST`, registers the opcode, and adds `opcodeDigest` backed by the existing SHA-256/SHA-1/RIPEMD-160 implementations plus `golang.org/x/crypto/sha3` for Keccak-256 and SHA3-256.
- `opcode_test.go`: adds known-answer vectors for all supported hash types over `0102` and empty input, keeps the Keccak-vs-SHA3 distinction pinned, and covers underflow, unknown selectors, non-minimal selectors, and oversized selectors.
- `README.md`: documents `OP_DIGEST` in the Cryptography opcode table.

### Notes
- No opcode-specific compute limit is added. Inputs remain bounded by the existing stack element size cap, matching the cost profile of the existing one-shot hash opcodes.
- `0xc3` was originally proposed in this PR as `OP_KECCAK256`; this revision generalizes that slot before merge rather than adding separate one-off hash opcodes.

### CODE REVIEW

**Comment by Kukks** [comment]
Let's try make this a multi style and repurpose op sha256? If there is a special byte in front we treat as keccak? Wdyt?

**Review by ghost — APPROVED** [review]
## ✅ Arkana Code Review — OP_KECCAK256

Clean, minimal addition. Reviewed the full diff and surrounding context.

### Verified

- **Correct hash primitive**: `sha3.NewLegacyKeccak256()` produces Ethereum-compatible Keccak-256 (legacy `0x01` padding), NOT NIST SHA3-256. Confirmed by cross-checking test vectors — both `0102` and empty-input digests are distinct from `sha3.New256()` outputs.
- **Test vectors canonical**: Empty-input hash `c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470` matches the well-known Ethereum Keccak-256 empty hash. The `0102` vector `22ae6da6...` also checks out.
- **Implementation mirrors `opcodeSha256`** (`opcode.go:1714`): pop → hash → push. Same error handling pattern. No divergence.
- **No compute budget concern**: No budget system exists in the codebase. Input bounded by the 520-byte element cap enforced at the stack level.
- **Opcode slot `0xc3` (195)**: Was `OP_UNKNOWN195` / `opcodeInvalid`. No downstream consumers — grepped all repos including ts-sdk, introspector-review. No breakage.
- **Test coverage**: `hashSpec` known-answer test + dedicated empty-input edge case + fuzz harness coverage via default case builder. Underflow (empty stack) covered by `hashSpec` internals.

### Cross-repo impact

None. Opcode 195 was previously invalid/unused. No SDK, client, or sister repo references this slot.

### Not protocol-critical

This is a new cryptographic primitive opcode (hashing only, no signing/spending logic). Does not touch VTXOs, forfeit paths, round lifecycle, or exit paths. No human sign-off required.

LGTM. Ship it.

**Review by ghost — APPROVED** [review]
## ✅ Arkana Code Review — OP_KECCAK256

Clean, minimal addition. Reviewed the full diff and surrounding context.

### Verified

- **Correct hash primitive**: `sha3.NewLegacyKeccak256()` produces Ethereum-compatible Keccak-256 (legacy `0x01` padding), NOT NIST SHA3-256. Confirmed by cross-checking test vectors — both `0102` and empty-input digests are distinct from `sha3.New256()` outputs.
- **Test vectors canonical**: Empty-input hash `c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470` matches the well-known Ethereum Keccak-256 empty hash. The `0102` vector `22ae6da6...` also checks out.
- **Implementation mirrors `opcodeSha256`** (`opcode.go:1714`): pop → hash → push. Same error handling pattern. No divergence.
- **No compute budget concern**: No budget system exists in the codebase. Input bounded by the 520-byte element cap enforced at the stack level.
- **Opcode slot `0xc3` (195)**: Was `OP_UNKNOWN195` / `opcodeInvalid`. No downstream consumers — grepped all repos including ts-sdk, introspector-review. No breakage.
- **Test coverage**: `hashSpec` known-answer test + dedicated empty-input edge case + fuzz harness coverage via default case builder. Underflow (empty stack) covered by `hashSpec` internals.

### Cross-repo impact

None. Opcode 195 was previously invalid/unused. No SDK, client, or sister repo references this slot.

### Not protocol-critical

This is a new cryptographic primitive opcode (hashing only, no signing/spending logic). Does not touch VTXOs, forfeit paths, round lifecycle, or exit paths. No human sign-off required.

LGTM. Ship it.

**Comment by msinkec** [comment]
> Let's try make this a multi style and repurpose op sha256? If there is a special byte in front we treat as keccak? Wdyt?

We could only do this cleany by also having sha256 be selected by a special hash type byte so the first byte will always be the hash type selector.  But then we break compatibility and symetry to L1. At this point it'd also make more to make the opcode accept a seperate hash type input instead of embedding this info in a single blob.
I'd keep OP_SHA256 as is and maybe introduce a new OP_DIGEST or something like that:

`<hash_type> <data> OP_DIGEST -> <hash>`

and we could make it support different types of new hashes:

```
HashKeccak256 = 1
HashBlake2b256 = 2
HashBlake3 = 3
HashSha3_256 = 4
```

**Review by ghost — APPROVED** [review]
## ✅ Arkana Code Review — OP_KECCAK256 → OP_DIGEST (incremental)

Reviewed the new commit `extend to general OP_DIGEST`. Clean evolution from single-hash opcode to multi-hash dispatcher, following Kukks' suggestion.

### Verified (new commit only)

- **Dispatch logic correct**: `PopInt` for selector, `PopByteArray` for data. Switch covers all 5 hash types with correct implementations. Unknown/zero types fall through to `default` → fail. Matches the stack doc: `[... data hash_type] → [... digest]`.
- **SHA-256 via OP_DIGEST matches OP_SHA256**: Both use `sha256.Sum256`. The `"0102"` vector `a12871fee210fb8619291eaea194581cbd2531e4b23759d225f6806923f63222` matches the existing `OP_SHA256` hashSpec vector. Consistent.
- **Keccak-256 / SHA3-256 distinction preserved**: Legacy Keccak (`NewLegacyKeccak256`) and NIST SHA3 (`New256`) produce different outputs for the same input. Both empty-input vectors are canonical. The test suite pins this.
- **Test coverage thorough**: 10 valid vectors (5 algorithms × 2 inputs), 6 invalid vectors covering underflow, unknown type, non-minimal encoding, and oversized selector. `digestSpec` correctly expects `beforeDepth-1 == afterDepth` (2 pops, 1 push).
- **Fuzz harness**: OP_DIGEST inherits coverage from the default case builder since it has a spec entry.

### One nit (non-blocking)

**`opcode_test.go:1712-1717` — `hashSpec` error relaxation is unnecessary:**
```go
requireScriptErrorCodeIn(t, c.execErr,
    txscript.ErrInvalidStackOperation,
    txscript.ErrMinimalData,       // ← new
    txscript.ErrNumberTooBig,      // ← new
)
```
`hashSpec` is used by OP_SHA256, OP_SHA1, OP_RIPEMD160, OP_HASH160, OP_HASH256 — all of which only call `PopByteArray`, never `PopInt`. They should never produce `ErrMinimalData` or `ErrNumberTooBig`. This relaxation could mask a future regression in those opcodes. OP_DIGEST already has its own `digestSpec` with the correct error set. Consider reverting this hunk to keep the existing opcodes' error assertions tight.

### Not flagged

- Missing negative hash-type test (`scriptNum(-1)`) — would hit `default` correctly, just no explicit vector. Ultra-minor.
- No cross-repo consumers of `OP_KECCAK256` existed (grepped all repos), so the rename to `OP_DIGEST` breaks nothing.
- Not protocol-critical (hashing primitive, no signing/spending/VTXO logic).

LGTM. Ship it.

**Review by ghost — APPROVED** [review]
## ✅ Arkana Code Review — OP_KECCAK256 → OP_DIGEST (incremental)

Reviewed the new commit `extend to general OP_DIGEST`. Clean evolution from single-hash opcode to multi-hash dispatcher, following Kukks' suggestion.

### Verified (new commit only)

- **Dispatch logic correct**: `PopInt` for selector, `PopByteArray` for data. Switch covers all 5 hash types with correct implementations. Unknown/zero types fall through to `default` → fail. Matches the stack doc: `[... data hash_type] → [... digest]`.
- **SHA-256 via OP_DIGEST matches OP_SHA256**: Both use `sha256.Sum256`. The `"0102"` vector `a12871fee210fb8619291eaea194581cbd2531e4b23759d225f6806923f63222` matches the existing `OP_SHA256` hashSpec vector. Consistent.
- **Keccak-256 / SHA3-256 distinction preserved**: Legacy Keccak (`NewLegacyKeccak256`) and NIST SHA3 (`New256`) produce different outputs for the same input. Both empty-input vectors are canonical. The test suite pins this.
- **Test coverage thorough**: 10 valid vectors (5 algorithms × 2 inputs), 6 invalid vectors covering underflow, unknown type, non-minimal encoding, and oversized selector. `digestSpec` correctly expects `beforeDepth-1 == afterDepth` (2 pops, 1 push).
- **Fuzz harness**: OP_DIGEST inherits coverage from the default case builder since it has a spec entry.

### One nit (non-blocking)

**`opcode_test.go:1712-1717` — `hashSpec` error relaxation is unnecessary:**
```go
requireScriptErrorCodeIn(t, c.execErr,
    txscript.ErrInvalidStackOperation,
    txscript.ErrMinimalData,       // ← new
    txscript.ErrNumberTooBig,      // ← new
)
```
`hashSpec` is used by OP_SHA256, OP_SHA1, OP_RIPEMD160, OP_HASH160, OP_HASH256 — all of which only call `PopByteArray`, never `PopInt`. They should never produce `ErrMinimalData` or `ErrNumberTooBig`. This relaxation could mask a future regression in those opcodes. OP_DIGEST already has its own `digestSpec` with the correct error set. Consider reverting this hunk to keep the existing opcodes' error assertions tight.

### Not flagged

- Missing negative hash-type test (`scriptNum(-1)`) — would hit `default` correctly, just no explicit vector. Ultra-minor.
- No cross-repo consumers of `OP_KECCAK256` existed (grepped all repos), so the rename to `OP_DIGEST` breaks nothing.
- Not protocol-critical (hashing primitive, no signing/spending/VTXO logic).

LGTM. Ship it.

File: pkg/arkade/opcode.go [hunk 1]
```diff
@@ -1722,6 +1723,57 @@ func opcodeSha256(op *opcode, data []byte, vm *Engine) error {
 	return nil
 }
 
+// Hash type selectors for OP_DIGEST.
+const (
+	digestSHA256    = 1
+	digestSHA1      = 2
```
Comment (louisinger) [inline pkg/arkade/opcode.go:1729]: nit: isn't SHA1 considered unsafe ?

File: pkg/arkade/opcode.go [hunk 2]
```diff
@@ -1722,6 +1723,57 @@ func opcodeSha256(op *opcode, data []byte, vm *Engine) error {
 	return nil
 }
 
+// Hash type selectors for OP_DIGEST.
+const (
+	digestSHA256    = 1
```
Comment (louisinger) [inline pkg/arkade/opcode.go:1728]: if we have SHA256 and RIPEMD160, let's include HASH160 also ?

File: pkg/arkade/opcode_test.go [hunk 3]
```diff
@@ -1730,6 +1734,119 @@ func hashSpec(op byte, inputHex, expectedHex string, hashLen int) *opcodeSpec {
 	}
 }
 
+// digestVector builds an OP_DIGEST valid vector hashing inputHex under the given
+// hash-type selector and asserting the expectedHex digest. The stack layout is
+// [data hash_type] (selector on top), matching the opcode's pop order.
+func digestVector(name string, hashType int, inputHex, expectedHex string) opcodeVector {
```
Comment (louisinger) [inline pkg/arkade/opcode_test.go]: let's hardcode vectors (1 per hash function is enough)

File: pkg/arkade/opcode.go (hunk 1 above)
Comment (msinkec) [reply pkg/arkade/opcode.go:1729]: There are collisions but for preimage checks its still fine to use.

File: pkg/arkade/opcode.go (hunk 2 above)
Comment (msinkec) [reply pkg/arkade/opcode.go:1728]: I've chose not to include double hash funcs, since you can still combine them explicitly.

File: pkg/arkade/opcode.go [hunk 4]
```diff
@@ -1722,6 +1723,57 @@ func opcodeSha256(op *opcode, data []byte, vm *Engine) error {
...
+	case digestSHA3_256:
+		h := sha3.New256()
+		h.Write(buf)
+		digest = h.Sum(nil)
+	default:
+		str := fmt.Sprintf("unsupported OP_DIGEST hash type %d", hashType)
+		return scriptError(txscript.ErrInvalidStackOperation, str)
+	}
+
+	vm.dstack.PushByteArray(digest)
```
Comment (chris-ricketts) [inline pkg/arkade/opcode.go]: Nit: re-use `calcHash` for brevity:

```suggestion
	switch hashType {
	case digestSHA256:
		h := sha256.Sum256(buf)
		vm.dstack.PushByteArray(h[:])
	case digestSHA1:
		h := sha1.Sum(buf)
		vm.dstack.PushByteArray(h[:])
	case digestRIPEMD160:
		vm.dstack.PushByteArray(calcHash(buf, ripemd160.New()))
	case digestKeccak256:
		vm.dstack.PushByteArray(calcHash(buf, sha3.NewLegacyKeccak256()))
	case digestSHA3_256:
		vm.dstack.PushByteArray(calcHash(buf, sha3.New256()))
	default:
		return scriptError(txscript.ErrInvalidStackOperation,
			fmt.Sprintf("unsupported OP_DIGEST hash type %d", hashType))
	}
```

**Review by chris-ricketts — APPROVED** [review]
One nit but otherwise LGTM

**Review by ghost — APPROVED** [review]
## ✅ Arkana Code Review — OP_DIGEST (incremental, 2 new commits)

Reviewed `simplify digest test` (90e7e66) and `Update pkg/arkade/opcode.go` (a3eaf36).

### opcode.go — push-per-branch refactor ✅

Inlining `PushByteArray` into each case arm instead of collecting into `var digest` is cleaner. Using `calcHash` for Keccak-256 and SHA3-256 (matching RIPEMD-160) is a good consistency win — no behavioral change since `calcHash` does the same `Write`+`Sum(nil)`.

### opcode_test.go — test simplification ✅

Removing the `digestVector` helper and using inline `[][]byte` literals with `mustDecodeHex` is consistent with how other test specs are structured. Vectors are identical — same hex, same expected outputs.

**One note:** the 5 empty-input vectors were dropped. Not a regression risk (the non-empty vectors still pin all 5 algorithms and the Keccak/SHA3 distinction), but worth being aware of. Empty-input hashes are well-known constants unlikely to break.

### Previous nit still open

The `hashSpec` error relaxation at `opcode_test.go:1715` (adding `ErrMinimalData` and `ErrNumberTooBig` to the existing hash opcodes' error set) is still present. Still non-blocking — just loosens assertions for opcodes that can't produce those errors. Mentioning for completeness; doesn't gate merge.

### Verdict

Both commits are pure simplification with no behavioral change. Still LGTM.

**Review by ghost — APPROVED** [review]
## ✅ Arkana Code Review — OP_DIGEST (incremental, 2 new commits)

Reviewed `simplify digest test` (90e7e66) and `Update pkg/arkade/opcode.go` (a3eaf36).

### opcode.go — push-per-branch refactor ✅

Inlining `PushByteArray` into each case arm instead of collecting into `var digest` is cleaner. Using `calcHash` for Keccak-256 and SHA3-256 (matching RIPEMD-160) is a good consistency win — no behavioral change since `calcHash` does the same `Write`+`Sum(nil)`.

### opcode_test.go — test simplification ✅

Removing the `digestVector` helper and using inline `[][]byte` literals with `mustDecodeHex` is consistent with how other test specs are structured. Vectors are identical — same hex, same expected outputs.

**One note:** the 5 empty-input vectors were dropped. Not a regression risk (the non-empty vectors still pin all 5 algorithms and the Keccak/SHA3 distinction), but worth being aware of. Empty-input hashes are well-known constants unlikely to break.

### Previous nit still open

The `hashSpec` error relaxation at `opcode_test.go:1715` (adding `ErrMinimalData` and `ErrNumberTooBig` to the existing hash opcodes' error set) is still present. Still non-blocking — just loosens assertions for opcodes that can't produce those errors. Mentioning for completeness; doesn't gate merge.

### Verdict

Both commits are pure simplification with no behavioral change. Still LGTM.

**Review by louisinger — APPROVED** [review]
